domain/miningmanager/mempool/mempool.go

```diff
@@ -1,36 +1,900 @@
+
+func (mp *mempool) updateBlockTransactionChainedTransactions(tx *consensusexternalapi.DomainTransaction) {
+	prevOut := consensusexternalapi.DomainOutpoint{TransactionID: *hashserialization.TransactionID(tx)}
+	for txOutIdx := range tx.Outputs {
+		// Skip to the next available output if there are none.
+		prevOut.Index = uint32(txOutIdx)
+		txDesc, exists := mp.chainedTransactionByPreviousOutpoint[prevOut]
+		if !exists {
+			continue
+		}
+
+		txDesc.depCount--
+		// If the transaction is not chained anymore, move it into the main pool
+		if txDesc.depCount == 0 {
+			// Transaction may be already removed by recursive calls, if removeRedeemers is true.
+			// So avoid moving it into main pool
+			txDescID := hashserialization.TransactionID(txDesc.DomainTransaction)
+			if _, ok := mp.chainedTransactions[*txDescID]; ok {
+				delete(mp.chainedTransactions, *txDescID)
+				mp.pool[*txDescID] = txDesc
+			}
+		}
+		delete(mp.chainedTransactionByPreviousOutpoint, prevOut)
+	}
+}
+
+// removeChainTransaction removes a chain transaction and all of its relation as a result of double spend.
+//
+// This function MUST be called with the mempool lock held (for writes).
+func (mp *mempool) removeChainTransaction(tx *consensusexternalapi.DomainTransaction) {
+	delete(mp.chainedTransactions, *hashserialization.TransactionID(tx))
+	for _, txIn := range tx.Inputs {
+		delete(mp.chainedTransactionByPreviousOutpoint, txIn.PreviousOutpoint)
+	}
+}
+
+// removeDoubleSpends removes all transactions which spend outputs spent by the
+// passed transaction from the memory pool. Removing those transactions then
+// leads to removing all transactions which rely on them, recursively. This is
+// necessary when a block is connected to the DAG because the block may
+// contain transactions which were previously unknown to the memory pool.
+//
+// This function MUST be called with the mempool lock held (for writes).
+func (mp *mempool) removeDoubleSpends(tx *consensusexternalapi.DomainTransaction) error {
+	txID := *hashserialization.TransactionID(tx)
+	for _, txIn := range tx.Inputs {
+		if txRedeemer, ok := mp.mempoolUTXOSet.poolTransactionBySpendingOutpoint(txIn.PreviousOutpoint); ok {
+			if !(*hashserialization.TransactionID(txRedeemer) == txID) {
+				err := mp.removeTransactionAndItsChainedTransactions(txRedeemer)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// addTransaction adds the passed transaction to the memory pool. It should
+// not be called directly as it doesn't perform any validation. This is a
+// helper for maybeAcceptTransaction.
+//
+// This function MUST be called with the mempool lock held (for writes).
+func (mp *mempool) addTransaction(tx *consensusexternalapi.DomainTransaction, mass uint64, fee uint64, parentsInPool []consensusexternalapi.DomainOutpoint) (*txDescriptor, error) {
+	// Add the transaction to the pool and mark the referenced outpoints
+	// as spent by the pool.
+	txD := &txDescriptor{
```

Comment:
I actually didn't mind txDesc too much (well, I did, but no commenting on names and everything), and what really bothered me was txD.

I didn't know that tx's had a D :stuck_out_tongue_closed_eyes: . Sorry not sorry.

Reply:
looool